File: lib/data/repository/model/CoffeeResponse.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:technical_test/domain/coffee/model/CoffeeDomainModel.dart';

part 'CoffeeResponse.freezed.dart';

@freezed
class CoffeeHotResponse with _$CoffeeHotResponse {
  const factory CoffeeHotResponse({
    int? id,
    String? title,
    List<String>? ingredients,
    String? image,
    String? description,
  }) = _CoffeeHotResponse;

  factory CoffeeHotResponse.fromJson(Map<String, dynamic> json) {
    final rawIngredients = json['ingredients'];
    List<String> parsedIngredients;

    if (rawIngredients is String) {
      parsedIngredients = rawIngredients
          .split(',')
          .map((e) => e.trim())
          .where((e) => e.isNotEmpty)
          .toList();
    } else if (rawIngredients is List) {
      parsedIngredients = rawIngredients
          .map((e) => e.toString().trim())
          .where((e) => e.isNotEmpty)
          .toList();
    } else {
      parsedIngredients = [];
    }

    return CoffeeHotResponse(
      id: json['id'] as int?,
      title: json['title'] as String?,
      description: json['description'] as String?,
      image: json['image'] as String?,
      ingredients: parsedIngredients,
    );
  }
}

extension CoffeeHotListMapper on List<CoffeeHotResponse> {
  List<CoffeeDomainModel> toDomainList() {
    return map((e) => e.toDomain()).toList();
  }
}

extension CoffeeHotMapper on CoffeeHotResponse {
  CoffeeDomainModel toDomain() {
    return CoffeeDomainModel(
      id: id ?? 0,
      description: description ?? '',
      image: image ?? '',
      ingredients: ingredients ?? [],
      title: title ?? '',
    );
  }
}

File: app/src/main/java/com/example/coffeepedia/data/repository/model/CoffeeResponse.kt
package com.example.coffeepedia.data.repository.model

import com.example.coffeepedia.domain.model.CoffeeDomainModel

data class CoffeeResponse(
    val title: String?,
    val description: String?,
    val ingredients: Any?,
    val image: Any?,
    val id: String?,
){
    companion object {
        fun CoffeeResponse.toDomain(): CoffeeDomainModel {
            val parsedIngredients: List<String> = when (ingredients) {
                is String -> {
                    ingredients
                        .split(",")
                        .map { it.trim() }
                        .filter { it.isNotEmpty() }
                }

                is List<*> -> {
                    ingredients
                        .mapNotNull { it?.toString()?.trim() }
                        .filter { it.isNotEmpty() }
                }

                else -> emptyList()
            }

            val parseImages: String = when (image) {
                is String -> {
                    image
                }

                is List<*> -> {
                    image.elementAt(0).toString()
                }

                else -> ""
            }

            return CoffeeDomainModel(
                id = id.orEmpty(),
                title = title.orEmpty(),
                description = description.orEmpty(),
                image = parseImages,
                ingredients = parsedIngredients
            )
        }

        fun List<CoffeeResponse>.toDomainList(): List<CoffeeDomainModel> =
            map { it.toDomain() }
    }
}
